File: app/src/testSocial/java/com/worldventures/dreamtrips/social/BucketInteractorBaseTest.java
package com.worldventures.dreamtrips.social;

import android.support.annotation.CallSuper;

import com.techery.spares.session.SessionHolder;
import com.techery.spares.storage.complex_objects.Optional;
import com.worldventures.dreamtrips.core.janet.cache.CacheResultWrapper;
import com.worldventures.dreamtrips.core.janet.cache.storage.ActionStorage;
import com.worldventures.dreamtrips.core.janet.cache.storage.MemoryStorage;
import com.worldventures.dreamtrips.core.repository.SnappyRepository;
import com.worldventures.dreamtrips.core.session.UserSession;
import com.worldventures.dreamtrips.core.test.BaseTest;
import com.worldventures.dreamtrips.core.test.MockDaggerActionService;
import com.worldventures.dreamtrips.core.test.MockHttpActionService;
import com.worldventures.dreamtrips.core.test.StubServiceWrapper;
import com.worldventures.dreamtrips.modules.bucketlist.model.BucketItem;
import com.worldventures.dreamtrips.modules.bucketlist.service.BucketInteractor;
import com.worldventures.dreamtrips.modules.bucketlist.service.storage.BucketListDiskStorage;
import com.worldventures.dreamtrips.modules.common.model.User;

import org.junit.Before;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.mockito.Spy;

import java.io.IOException;
import java.net.URISyntaxException;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import io.techery.janet.CommandActionService;
import io.techery.janet.Janet;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public abstract class BucketInteractorBaseTest extends BaseTest {
    static final int MOCK_USER_ID = 1;

    @Mock
    SessionHolder<UserSession> mockSessionHolder;
    @Spy
    MemoryStorage<List<BucketItem>> mockMemoryStorage;

    @Mock
    SnappyRepository mockDb;

    Janet janet;

    StubServiceWrapper httpStubWrapper;

    BucketInteractor bucketInteractor;

    UserSession userSession;

    private MockDaggerActionService daggerActionService;

    @CallSuper
    @Before
    public void setup() throws URISyntaxException, IOException {
        MockitoAnnotations.initMocks(this);

        CacheResultWrapper cacheResultWrapper = cachedService(
                daggerActionService = new MockDaggerActionService(new CommandActionService()));
        for (ActionStorage storage : storageSet()) {
            cacheResultWrapper.bindStorage(storage.getActionClass(), storage);
        }

        janet = new Janet.Builder()
                .addService(cacheResultWrapper)
                .addService(cachedService(
                        httpStubWrapper = new StubServiceWrapper(mockHttpService()))
                )
                .build();

        daggerActionService.registerProvider(Janet.class, () -> janet);
        daggerActionService.registerProvider(SnappyRepository.class, () -> mockDb);
        daggerActionService.registerProvider(SessionHolder.class, () -> mockSessionHolder);
        daggerActionService.registerProvider(BucketInteractor.class, () -> bucketInteractor);

        bucketInteractor = new BucketInteractor(janet);

        userSession = mock(UserSession.class);
        User mockUser = mock(User.class);

        when(mockUser.getId()).thenReturn(MOCK_USER_ID);
        when(userSession.getUser()).thenReturn(mockUser);
        when(mockSessionHolder.get()).thenReturn(Optional.of(userSession));
    }

    protected abstract MockHttpActionService mockHttpService();

    MockDaggerActionService daggerActionService() {
        return daggerActionService;
    }

    protected Set<ActionStorage> storageSet() {
        Set<ActionStorage> storageSet = new HashSet<>();
        storageSet.add(new BucketListDiskStorage(mockMemoryStorage, mockDb, mockSessionHolder));

        return storageSet;
    }
}

File: app/src/testSocial/java/com/worldventures/dreamtrips/social/bucket/junit/BucketInteractorBaseTest.kt
package com.worldventures.dreamtrips.social.bucket.junit

import android.support.annotation.CallSuper
import com.techery.spares.session.SessionHolder
import com.techery.spares.storage.complex_objects.Optional
import com.worldventures.dreamtrips.BaseTest
import com.worldventures.dreamtrips.core.janet.cache.storage.ActionStorage
import com.worldventures.dreamtrips.core.repository.SnappyRepository
import com.worldventures.dreamtrips.core.session.UserSession
import com.worldventures.dreamtrips.janet.MockDaggerActionService
import com.worldventures.dreamtrips.janet.StubServiceWrapper
import com.worldventures.dreamtrips.modules.bucketlist.service.BucketInteractor
import com.worldventures.dreamtrips.modules.bucketlist.service.storage.BucketListDiskStorage
import com.worldventures.dreamtrips.modules.bucketlist.service.storage.BucketMemoryStorage
import com.worldventures.dreamtrips.modules.common.model.User
import com.worldventures.dreamtrips.modules.infopages.StaticPageProvider
import io.techery.janet.CommandActionService
import io.techery.janet.Janet
import io.techery.janet.http.test.MockHttpActionService
import org.junit.Before
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.Mockito.mock
import org.mockito.MockitoAnnotations
import org.mockito.Spy
import java.util.*

abstract class BucketInteractorBaseTest : BaseTest() {
    init {
        MockitoAnnotations.initMocks(this)
    }

    @Mock var mockSessionHolder: SessionHolder<UserSession>? = null
    @Mock var mockDb: SnappyRepository? = null
    @Mock var staticPageProvider: StaticPageProvider? = null

    @Spy var mockMemoryStorage: BucketMemoryStorage? = null

    var janet: Janet? = null

    val httpStubWrapper: StubServiceWrapper by lazy {
        StubServiceWrapper(mockHttpService())
    }

    var bucketInteractor: BucketInteractor? = null

    var userSession: UserSession? = null

    lateinit var daggerActionService: MockDaggerActionService

    @CallSuper
    @Before
    open fun setup() {
        val cacheResultWrapper = cachedService(CommandActionService())
        for (storage in storageSet()) {
            cacheResultWrapper.bindStorage(storage.actionClass, storage)
        }
        daggerActionService = MockDaggerActionService(cacheResultWrapper)
        daggerActionService.registerProvider(Janet::class.java) { janet }
        daggerActionService.registerProvider(SnappyRepository::class.java) { mockDb }
        daggerActionService.registerProvider(SessionHolder::class.java) { mockSessionHolder }
        daggerActionService.registerProvider(BucketInteractor::class.java) { bucketInteractor }
        daggerActionService.registerProvider(StaticPageProvider::class.java, { staticPageProvider })

        janet = Janet.Builder()
                .addService(daggerActionService)
                .addService(cachedService(httpStubWrapper))
                .build()

        bucketInteractor = BucketInteractor(janet)

        userSession = mock(UserSession::class.java)
        val mockUser = mock(User::class.java)

        `when`(mockUser.id).thenReturn(MOCK_USER_ID)
        `when`(userSession!!.user).thenReturn(mockUser)
        `when`(mockSessionHolder!!.get()).thenReturn(Optional.of(userSession))
        `when`(staticPageProvider!!.uploaderyUrl).thenReturn("http://test-uploadery")
    }

    protected abstract fun mockHttpService(): MockHttpActionService

    internal fun daggerActionService(): MockDaggerActionService {
        return daggerActionService
    }

    protected open fun storageSet(): Set<ActionStorage<*>> {
        val storageSet = HashSet<ActionStorage<*>>()
        storageSet.add(BucketListDiskStorage(mockMemoryStorage, mockDb))

        return storageSet
    }

    companion object {
        val MOCK_USER_ID = 1
    }
}
